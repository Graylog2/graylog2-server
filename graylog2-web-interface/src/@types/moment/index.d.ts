/*
 * Copyright (C) 2020 Graylog, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the Server Side Public License, version 1,
 * as published by MongoDB, Inc.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * Server Side Public License for more details.
 *
 * You should have received a copy of the Server Side Public License
 * along with this program. If not, see
 * <http://www.mongodb.com/licensing/server-side-public-license>.
 */

// Type definitions for moment-precise-range 0.2
// Original:
// https://github.com/DefinitelyTyped/DefinitelyTyped/blob/bc1eb3b28e41b2ed54865aecf134eae6c0b23add/types/moment-precise-range-plugin/index.d.ts
// Project: https://github.com/codebox/moment-precise-range
// Definitions by: Mitchell Grice <https://github.com/gricey432>
// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped

import moment = require('moment');

export = moment;

declare module 'moment' {
  interface PreciseRangeValueObject {
    years: number;
    months: number;
    days: number;
    hours: number;
    minutes: number;
    seconds: number;
    firstDateWasLater: boolean;
  }

  interface Moment {
    preciseDiff(d2: Moment, returnValueObject?: false): string;
    preciseDiff(d2: Moment, returnValueObject: true): PreciseRangeValueObject;
  }

  function preciseDiff(d1: Moment, d2: Moment, returnValueObject?: false): string;
  function preciseDiff(d1: Moment, d2: Moment, returnValueObject: true): PreciseRangeValueObject;
}
