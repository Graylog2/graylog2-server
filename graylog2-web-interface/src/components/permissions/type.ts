declare module 'graylog-web-plugin/plugin' {

  export interface PluginExports {
    'components.collection'?: {
      AddCollectionFormGroup: any,
    },
  }
}
