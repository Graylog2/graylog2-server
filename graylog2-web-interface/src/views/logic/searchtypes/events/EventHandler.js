// @flow strict
import { groupBy } from 'lodash';
import WidgetFormattingSettings from 'views/logic/aggregationbuilder/WidgetFormattingSettings';
import type { ChartDefinition } from 'views/components/visualizations/ChartData';

export type Event = {
  id: string,
  timestamp: string,
  message: string,
  alert: boolean,
  streams: Array<string>,
};

export type Events = Array<Event>;

type GroupedEvents = { [string]: Events };

type Shape = {
  type: 'line',
  y0: number,
  y1: number,
  x0: string,
  x1: string,
  opacity: number,
  line: {
    color: string,
  }
};

export type Shapes = Array<Shape>;
const eventsDisplayName = 'Alerts';
const defaultColor = '#d3d3d3';

export default {
  convert(events: Array<Event>) {
    return events;
  },

  toVisualizationData(events: Events = [],
    formattingSettings: WidgetFormattingSettings = WidgetFormattingSettings.create({})): { eventChartData: ChartDefinition, shapes: Shapes } {
    const groupedEvents: GroupedEvents = groupBy(events, e => e.timestamp);
    return {
      eventChartData: this.toChartData(groupedEvents, formattingSettings),
      shapes: this.toShapeData(Object.keys(groupedEvents), formattingSettings),
    };
  },

  transformGroupedEvents(events: GroupedEvents): [[string, Event | number]] {
    // $FlowFixMe Object.entries has only mixed values.
    return Object.entries(events).map(([timestamp:string, eventArray:Events]) => {
      if (!(eventArray instanceof Array)) {
        throw new Error('Unexpected data type');
      }

      if (eventArray.length > 1) {
        return [timestamp, eventArray.length];
      }

      return [timestamp, eventArray[0]];
    });
  },

  toChartData(events: GroupedEvents, formattingSettings: WidgetFormattingSettings): ChartDefinition {
    const { chartColors } = formattingSettings;
    const chartColor = chartColors[eventsDisplayName] || defaultColor;
    const values = this.transformGroupedEvents(events);
    const xValues: Array<string> = values.map(v => v[0]);
    const textValues: Array<string> = values.map((e) => {
      if (Object.prototype.hasOwnProperty.call(e[1], 'message')) {
        return e[1].message;
      }
      return `${e[1]} alerts occurred.`;
    });
    const yValues: Array<number> = values.map(() => 0);
    return {
      hovertemplate: '%{text}',
      mode: 'markers',
      name: eventsDisplayName,
      type: 'scatter',
      opacity: 0.5,
      x: xValues,
      y: yValues,
      text: textValues,
      marker: {
        size: 5,
        color: chartColor,
      },
    };
  },

  toShapeData(timestamps: Array<string>, formattingSettings: WidgetFormattingSettings = WidgetFormattingSettings.create({})): Shapes {
    const { chartColors } = formattingSettings;
    const shapeColor = chartColors[eventsDisplayName] || defaultColor;
    return timestamps.map(timestamp => ({
      layer: 'below',
      type: 'line',
      yref: 'paper',
      y0: 0,
      y1: 1,
      x0: timestamp,
      x1: timestamp,
      opacity: 0.5,
      line: {
        color: shapeColor,
      },
    }));
  },
};
