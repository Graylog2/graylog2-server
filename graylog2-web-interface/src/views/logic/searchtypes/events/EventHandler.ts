/*
 * Copyright (C) 2020 Graylog, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the Server Side Public License, version 1,
 * as published by MongoDB, Inc.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * Server Side Public License for more details.
 *
 * You should have received a copy of the Server Side Public License
 * along with this program. If not, see
 * <http://www.mongodb.com/licensing/server-side-public-license>.
 */
import groupBy from 'lodash/groupBy';

import type { ChartDefinition } from 'views/components/visualizations/ChartData';

export type Event = {
  id: string,
  timestamp: string,
  message: string,
  alert: boolean,
  streams: Array<string>,
};

export type Events = Array<Event>;

export const EVENT_COLOR = '#d3d3d3';

type GroupedEvents = { [key: string]: Events };

type Shape = {
  type: 'line',
  y0: number,
  y1: number,
  x0: string,
  x1: string,
  opacity: number,
  line?: {
    color: string,
  },
};

export type Shapes = Array<Shape>;

export const eventsDisplayName = 'Alerts';

type TimeFormatter = (timestamp: string) => string;

export default {
  convert(events: Array<Event>) {
    return events;
  },

  toVisualizationData(events: Events, formatTime: TimeFormatter): { eventChartData: ChartDefinition, shapes: Shapes } {
    const groupedEvents: GroupedEvents = groupBy(events, (e) => e.timestamp);

    return {
      eventChartData: this.toChartData(groupedEvents, formatTime),
      shapes: this.toShapeData(Object.keys(groupedEvents), formatTime),
    };
  },

  transformGroupedEvents(events: GroupedEvents): [string, Event | number][] {
    return Object.entries(events).map(([timestamp, eventArray]) => {
      if (!(eventArray instanceof Array)) {
        throw new Error('Unexpected data type');
      }

      if (eventArray.length > 1) {
        return [timestamp, eventArray.length];
      }

      return [timestamp, eventArray[0]];
    });
  },

  toChartData(events: GroupedEvents, formatTimestamp: TimeFormatter): ChartDefinition {
    const values = this.transformGroupedEvents(events);
    const xValues: Array<string> = values.map((v) => formatTimestamp(v[0]));
    const textValues: Array<string> = values.map((e) => {
      if (typeof e[1] !== 'number' && 'message' in e[1]) {
        return e[1].message;
      }

      return `${e[1]} alerts occurred.`;
    });
    const yValues: Array<number> = values.map(() => 0);

    return {
      hovertemplate: '%{text}',
      mode: 'markers',
      name: eventsDisplayName,
      type: 'scatter',
      opacity: 0.5,
      x: xValues,
      y: yValues,
      text: textValues,
    };
  },

  toShapeData(timestamps: Array<string>, formatTimestamp: TimeFormatter): Shapes {
    return timestamps.map((timestamp) => {
      const formattedTimestamp = formatTimestamp(timestamp);

      return {
        layer: 'below',
        type: 'line',
        yref: 'paper',
        y0: 0,
        y1: 1,
        x0: formattedTimestamp,
        x1: formattedTimestamp,
        opacity: 0.5,
      };
    });
  },
};
