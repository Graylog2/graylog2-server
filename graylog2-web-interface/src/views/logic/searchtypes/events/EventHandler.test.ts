/*
 * Copyright (C) 2020 Graylog, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the Server Side Public License, version 1,
 * as published by MongoDB, Inc.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * Server Side Public License for more details.
 *
 * You should have received a copy of the Server Side Public License
 * along with this program. If not, see
 * <http://www.mongodb.com/licensing/server-side-public-license>.
 */
import groupBy from 'lodash/groupBy';
import moment from 'moment-timezone';

import type { Event } from './EventHandler';
import EventHandler from './EventHandler';

const formatTimestamp = (timezone: string) => (timestamp: string) => moment(timestamp).tz(timezone).toISOString(true);

const groupByTimestamp = ((es: Event[]) => groupBy(es, (e) => e.timestamp));

describe('EventHandler convert', () => {
  const event: Event = {
    alert: false,
    id: '01DSMMX5M4ER2MR02DDB2JS93W',
    message: 'This is a emergency. Please stay calm.',
    streams: ['5cdab2293d27467fbe9e8a72'],
    timestamp: '2019-11-14T08:53:35.000Z',
  };

  it('should convert events to chart data and shapes', () => {
    const result = EventHandler.toVisualizationData([event], formatTimestamp('UTC'));

    expect(result).toMatchSnapshot();
  });

  it('should convert events to char data', () => {
    const result = EventHandler.toChartData(groupByTimestamp([event]), formatTimestamp('UTC'));

    expect(result).toEqual({
      hovertemplate: '%{text}',
      mode: 'markers',
      opacity: 0.5,
      name: 'Alerts',
      type: 'scatter',
      x: ['2019-11-14T08:53:35.000+00:00'],
      y: [0],
      text: ['This is a emergency. Please stay calm.'],
    });
  });

  it('should convert events to char data and use new timezone', () => {
    const result = EventHandler.toChartData(groupByTimestamp([event]), formatTimestamp('CET'));

    expect(result).toEqual({
      hovertemplate: '%{text}',
      mode: 'markers',
      opacity: 0.5,
      name: 'Alerts',
      type: 'scatter',
      x: ['2019-11-14T09:53:35.000+01:00'],
      y: [0],
      text: ['This is a emergency. Please stay calm.'],
    });
  });

  it('should group duplicate events by timestamp and convert events to char data', () => {
    const result = EventHandler.toChartData(groupByTimestamp([event, event, event]), formatTimestamp('UTC'));

    expect(result).toEqual({
      hovertemplate: '%{text}',
      mode: 'markers',
      opacity: 0.5,
      name: 'Alerts',
      type: 'scatter',
      x: ['2019-11-14T08:53:35.000+00:00'],
      y: [0],
      text: ['3 alerts occurred.'],
    });
  });

  it('should keep the color set by the user', () => {
    const result = EventHandler.toChartData(groupByTimestamp([event]), formatTimestamp('UTC'));

    expect(result).toEqual({
      hovertemplate: '%{text}',
      mode: 'markers',
      opacity: 0.5,
      name: 'Alerts',
      type: 'scatter',
      x: ['2019-11-14T08:53:35.000+00:00'],
      y: [0],
      text: ['This is a emergency. Please stay calm.'],
    });
  });

  it('should convert events to shape data', () => {
    const result = EventHandler.toShapeData([event.timestamp], formatTimestamp('UTC'));

    expect(result[0]).toEqual({
      layer: 'below',
      type: 'line',
      yref: 'paper',
      y0: 0,
      y1: 1,
      x0: '2019-11-14T08:53:35.000+00:00',
      x1: '2019-11-14T08:53:35.000+00:00',
      opacity: 0.5,
    });
  });

  it('should convert events to shape data with new timezone', () => {
    const result = EventHandler.toShapeData([event.timestamp], formatTimestamp('CET'));

    expect(result[0]).toEqual({
      layer: 'below',
      type: 'line',
      yref: 'paper',
      y0: 0,
      y1: 1,
      x0: '2019-11-14T09:53:35.000+01:00',
      x1: '2019-11-14T09:53:35.000+01:00',
      opacity: 0.5,
    });
  });
});
