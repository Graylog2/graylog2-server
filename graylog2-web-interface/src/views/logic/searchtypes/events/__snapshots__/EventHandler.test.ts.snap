// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EventHandler convert should convert events to chart data and shapes 1`] = `
{
  "eventChartData": {
    "hovertemplate": "%{text}",
    "mode": "markers",
    "name": "Alerts",
    "opacity": 0.5,
    "text": [
      "This is a emergency. Please stay calm.",
    ],
    "type": "scatter",
    "x": [
      "2019-11-14T08:53:35.000+00:00",
    ],
    "y": [
      0,
    ],
  },
  "shapes": [
    {
      "layer": "below",
      "opacity": 0.5,
      "type": "line",
      "x0": "2019-11-14T08:53:35.000+00:00",
      "x1": "2019-11-14T08:53:35.000+00:00",
      "y0": 0,
      "y1": 1,
      "yref": "paper",
    },
  ],
}
`;
